File: 0101-0200/0154-find-minimum-in-rotated-sorted-array-ii.go
package main

import "fmt"

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func findMin(nums []int) int {
    lo, hi := 0, len(nums)
    ans := nums[0]
    for lo < hi {
        mid := lo + ((hi - lo) >> 1)
        if nums[lo] == nums[mid] {
            ans = min(ans, nums[lo])  // 必须有
            lo++
        } else {
            if nums[lo] < nums[mid] {
                ans = min(ans, nums[lo])  // 必须有
                lo = mid + 1
            } else {
                ans = min(ans, nums[mid])
                hi = mid
            }
        }
    }
    return ans
}

func main() {
    fmt.Println(findMin([]int{3, 5, 1}))
    // fmt.Println(findMin([]int{2,2,2,0,1}))
}

File: 0101-0200/0154-find-minimum-in-rotated-sorted-array-ii.rs
struct Solution {
}

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let (mut lo, mut hi) = (0_usize, nums.len() - 1);
        while lo < hi {
            // if lo < hi, mid < hi
            let mid = lo + (hi - lo) / 2;
            if nums[mid] == nums[hi] {
                hi -= 1;
            } else if nums[mid] < nums[hi] {
                hi = mid;
            } else {
                lo = mid + 1
            }
        }
        nums[lo]
    }
}

fn main() {
    assert_eq!(Solution::find_min(vec![1, 3, 5]), 1);
    assert_eq!(Solution::find_min(vec![2, 2, 2, 0, 1]), 0);
}
